### PR TITLE
Fix the find_xyz_cut_coord function

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Fixes:
- #1189: always returns a list
- #1190: always returns a point in real world coordinate or raise a meaningful error
- #1191: internally converts to float during masking (can be discussed later)
- #1193: force mask to be a niimg (I did not change the parameter name but maybe I should?)
- #1194: now works for binary images

I am still a bit doubtful about the "second threshold" and all the magic it does. Do we have a reference about that somewhere?

### CODE REVIEW

File: nilearn/plotting/find_cuts.py [hunk 1]
```diff
@@ -53,66 +54,72 @@ def find_xyz_cut_coords(img, mask=None, activation_threshold=None):
...
 
-    # To speed up computations, we work with partial views of the array,
-    # and keep track of the offset
-    offset = np.zeros(3)
+    # We have 3 potential masks here:
+    # - a mask provided by the user
+    # - the activation threshold
+    # - if data is a masked_array, the associated mask
+
+    masks = []
```
Comment (GaelVaroquaux) [inline nilearn/plotting/find_cuts.py]: Can we call this "mask_imgs"?

File: nilearn/plotting/find_cuts.py [hunk 2]
```diff
@@ -19,25 +19,26 @@
...
 
         Parameters
         -----------
         img : 3D Nifti1Image
             The brain map.
-        mask : 3D ndarray, boolean, optional
-            An optional brain mask.
-        activation_threshold : float, optional
-            The lower threshold to the positive activation. If None, the
+        mask : 3D Nifti1Image
```
Comment (GaelVaroquaux) [inline nilearn/plotting/find_cuts.py]: Can we call this mask_img?

File: nilearn/plotting/find_cuts.py [hunk 3]
```diff
@@ -53,66 +54,72 @@ def find_xyz_cut_coords(img, mask=None, activation_threshold=None):
...
-        my_map *= mask
-        offset += [slice_x.start, slice_y.start, slice_z.start]
+    if len(masks) > 0:
+        mask = intersect_masks(masks, threshold=1)
+        data = unmask(apply_mask([img], mask)[0], mask)
+    else:
+        # Get rid of potential memmapping
+        data = as_ndarray(data)
+        # XXX Is that necessary given _saf_get_data?
+        data = data.copy()
```
Comment (GaelVaroquaux) [inline nilearn/plotting/find_cuts.py]: I believe that the copy is necessary to get rid of the memmapping in a robust way. However, I am not sure how useful this is.

**Comment by GaelVaroquaux** [comment]
You have a bunch of failing tests. I'll let you look at them.

**Comment by AlexandreAbraham** [comment]
Question: there is a test that explicitely check that, in case of empty mask, the function returns the middle of the image. Is that a requirement of the function? Raising an error seems more natural to me.

**Comment by GaelVaroquaux** [comment]
> Raising an error seems more natural to me.

The error is a bad idea: think of an automated pipeline with plotting:
you don't want it to crash because of the plotting when the image is
fully sparse.

**Comment by AlexandreAbraham** [comment]
> The error is a bad idea: think of an automated pipeline with plotting: you don't want it to crash because of the plotting when the image is fully sparse.

I can think of that ;). Whether or not a plotting function should fail is up to the plotting function (altough, in that case, I prefer to have a special display rather than an empty image). The function `find_xyz_cut_coords` is supposed to find the center of mass of the largest component and nothing says that it returns "garbage" value if there is not.

I think that the problem lies in the ambiguity of the function. It is used to find cut coords in plotting functions (in that case, returning the center is OK) but people clearly uses it to find points of interest in their maps (in that case, returning the center is invalid).

File: nilearn/plotting/find_cuts.py [hunk 4]
```diff
@@ -53,66 +54,77 @@ def find_xyz_cut_coords(img, mask=None, activation_threshold=None):
...
+    non_zero = np.abs(data) > 0.
+    if activation_threshold == 'auto':
+        activation_threshold = fast_abs_percentile(
+            data[data != 0].ravel(), 80)
+        non_zero = np.abs(data) >= activation_threshold - 1.e-15
+        if non_zero.max() == 0:
+            raise ValueError('All voxels were masked by the auto threshold. '
+                             'Please disable it or provide a custom value.')
+
+    largest = largest_connected_component(non_zero)
```
Comment (banilo) [inline nilearn/plotting/find_cuts.py]: add comment what the slightly-hacky-looking code below does

File: nilearn/plotting/find_cuts.py [hunk 5]
```diff
@@ -53,66 +54,77 @@ def find_xyz_cut_coords(img, mask=None, activation_threshold=None):
...
+    slice_x, slice_y, slice_z = ndimage.find_objects(largest)[0]
+    data = data[slice_x, slice_y, slice_z]
+    largest = largest[slice_x, slice_y, slice_z]
+    data *= largest
 
     # For the second threshold, we use a mean, as it is much faster,
     # althought it is less robust
-    second_threshold = np.abs(np.mean(my_map[mask]))
-    second_mask = (np.abs(my_map) > second_threshold)
+    # XXX Can somebody explain what this code does?
```
Comment (banilo) [inline nilearn/plotting/find_cuts.py]: ;-)

**Comment by GaelVaroquaux** [comment]
The tests are still failing: you need to use sklearn.utils.testing.assert_raises_regexp" to be compatible with old versions of numpy.

**Comment by GaelVaroquaux** [comment]
> I think that the problem lies in the ambiguity of the function. It is used to
> find cut coords in plotting functions (in that case, returning the center is
> OK) but people clearly uses it to find points of interest in their maps (in
> that case, returning the center is invalid).

Do they really? What's the definition of a point of interest? If people
want peak detections, than this function is not what they should be
using.

I think that where I've seen it used it to find where to put a marker in
the center of a region, eg in a plot_connectome. That's also a
visualization usecase, hence acceptable to be lenient with input, and
blurry with output.

Beside, it's in the plotting sub-module, so I would expect such a
behavior.

We cannot merge a PR removing this behavior. This behavior is important
eg in QC pipelines that use nilearn for quick visualization. We would
break their usage.

File: nilearn/plotting/find_cuts.py (hunk 3 above)
Comment (AlexandreAbraham) [inline nilearn/plotting/find_cuts.py]: Let's leave it for now.

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.
